### ISSUE TITLE
Units handling inconsistency and bug

### ISSUE THREAD

**Issue by jmilloy** [creare-com/podpac#449]
**Description**
I think this is two issues.
 * `UnitsDataArray.__eq__` tries to convert units even if the RHS is just a basic float and throws an exception because the float is dimensionless
 *  `DataSource` ignores `units`?

**Steps to Reproduce**

Algorithm node that throws an exception on `==`:

```
>>> node = podpac.algorithm.Arange(units='psi')
>>> out = node.eval(podpac.Coordinates([[0, 1]], dims=['lat']))
>>> out
<xarray.UnitsDataArray (lat: 2)>
array([0., 1.])
Coordinates:
  * lat      (lat) float64 0.0 1.0
Attributes:
    layer_style:  <podpac.core.style.Style object at 0x7f64195b4910>
    crs:          +proj=longlat +datum=WGS84 +no_defs +vunits=m
    units:        pound_force_per_square_inch
>>> out[0] == 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jmilloy/Creare/Pipeline/podpac/podpac/core/units.py", line 507, in func
    multiplier = self._get_unit_multiplier(other)
  File "/home/jmilloy/Creare/Pipeline/podpac/podpac/core/units.py", line 78, in _get_unit_multiplier
    multiplier = otheru.to(myu.u).magnitude
  File "/home/jmilloy/Creare/Pipeline/_podpac-38_/lib/python3.8/site-packages/pint/quantity.py", line 661, in to
    magnitude = self._convert_magnitude_not_inplace(other, *contexts, **ctx_kwargs)
  File "/home/jmilloy/Creare/Pipeline/_podpac-38_/lib/python3.8/site-packages/pint/quantity.py", line 610, in _convert_magnitude_not_inplace
    return self._REGISTRY.convert(self._magnitude, self._units, other)
  File "/home/jmilloy/Creare/Pipeline/_podpac-38_/lib/python3.8/site-packages/pint/registry.py", line 954, in convert
    return self._convert(value, src, dst, inplace)
  File "/home/jmilloy/Creare/Pipeline/_podpac-38_/lib/python3.8/site-packages/pint/registry.py", line 1840, in _convert
    return super()._convert(value, src, dst, inplace)
  File "/home/jmilloy/Creare/Pipeline/_podpac-38_/lib/python3.8/site-packages/pint/registry.py", line 1442, in _convert
    return super()._convert(value, src, dst, inplace)
  File "/home/jmilloy/Creare/Pipeline/_podpac-38_/lib/python3.8/site-packages/pint/registry.py", line 987, in _convert
    raise DimensionalityError(src, dst, src_dim, dst_dim)
pint.errors.DimensionalityError: Cannot convert from 'dimensionless' (dimensionless) to 'pound_force_per_square_inch' ([mass] / [length] / [time] ** 2)
```

But no issue with a Datasource node, presumably because `out` contains no `units` attr:

```
>>> node = podpac.data.Array(source=[0, 1], coordinates=podpac.Coordinates([[0, 1]], dims=['lat']), units='psi')
>>> out = node.eval(node.coordinates)
>>> out
<xarray.UnitsDataArray (lat: 2)>
array([0., 1.])
Coordinates:
  * lat      (lat) float64 0.0 1.0
Attributes:
    layer_style:  <podpac.core.style.Style object at 0x7f6419c6aa90>
    crs:          +proj=longlat +datum=WGS84 +no_defs +vunits=m
>>> out[0] == 0
<xarray.UnitsDataArray ()>
array(True)
Coordinates:
    lat      float64 0.0
```

**Expected Behavior**

 * The units should be in the output UnitsDataArray in both cases.
 *  The `==` should work in both cases; I'm guessing that the UnitsDataArray needs to only convert units if the RHS is also a UnitsDataArray.

**Comment by jmilloy** [comment]
The UnitsDataArray is clearly trying to default to 1 here:

```
    def _get_unit_multiplier(self, other):
        multiplier = 1
        if (self.attrs.get("units", None) or getattr(other, "units", None)) and settings["ENABLE_UNITS"]:
            otheru = ureg.Quantity(1, getattr(other, "units", "1"))
            myu = ureg.Quantity(1, getattr(self, "units", "1"))
            multiplier = otheru.to(myu.u).magnitude
        return multiplier
```

**Comment by jmilloy** [comment]
@mpu-creare I think I need you to make a call on this. I would have thought we should only compute a multiplier if both `self` and `other` have units, but you clearly coded it to be if *either* have units.

Unfortunately, the `ureg.Quantity(1, 'getattr(other, "units", "1")) gives `dimensionless` units `<Quantity(1, 'dimensionless')>` if `other.units` are not defined, which then can't be converted.

Can we use

```
if getattr(self, "units", None) and getattr(other, "units", None) and settings['ENABLE_UNITS"]:
    ...
```

Alternatively, maybe we need to use the self.units for `otheru` and `other.units` for myu as the defaults?

```
def _get_unit_multiplier(self, other):
        multiplier = 1
        if (getattr(self, "units", None) or getattr(other, "units", None)) and settings["ENABLE_UNITS"]:
            otheru = ureg.Quantity(1, getattr(other, "units", getattr(self, "units")))
            myu = ureg.Quantity(1, getattr(self, "units", getattr(other, "units")))
            multiplier = otheru.to(myu.u).magnitude
        return multiplier
```

**Comment by jmilloy** [comment]
Sidenote, would be clearer if the if statement was the special case:

```
def _get_unit_multiplier(self, other):
    if not (getattr(self, "units", None) and getattr(other, "units", None) and settings["ENABLE_UNITS"]):
        return 1
    otheru = ureg.Quantity(1, other.units)
    myu = ureg.Quantity(1, self.units)
    return otheru.to(myu.u).magnitude
```

**Comment by jmilloy** [comment]
On the other front, `Datasource` works fine, maybe just something wrong with `Array`...

**Comment by jmilloy** [comment]
Ah, it's the InterpolationMixin. `ArrayRaw` works but not `Array`. Units need to propagate up.

**Comment by jmilloy** [comment]
This fixes that part:

```
class Interpolate(Node):
    ...
    @tl.default('units')
    def _use_source_units(self):
        return self.source.units
```

So just waiting on how you want to deal with the multiplier default. I see that the `self.units or other.units` pattern is used all over.

**Comment by mpu-creare** [comment]
@jmilloy the units are coded to be very strict. So when a quantity doesn't have units defined it assumed to be dimensionless. If that's a bad assumption, the error is thrown. So, for your example, it should probably be something like this:
```
out[0] == ureg.Quantity(0, out.attr['units'])
```
The whole reason why the "ENABLE_UNITS" setting exists is because the very strict unit enforcement can be a huge pain... it also doesn't work in all cases. Some algorithms strip units and there's nothing we can do about it. And you were right to propagate the units up through the InterpolationMixin -- that was a bug. 

Also, units is a very tentatively supported feature. We provided it to add value to lure folks to use PODPAC (and for our own sanity in some cases). 

Where are units being used now?

**Comment by jmilloy** [comment]
> Where are units being used now?

geowatch, should I just not use units? The units can just be in the `style`.

**Comment by mpu-creare** [comment]
I would recommend that approach for now -- or set "ENABLE_UNITS" to False. Otherwise you'll probably find and lot of bugs. :)

**Comment by jmilloy** [comment]
Okay, thanks. I'll push the Inerpolate fix and close this as wontfix (working as intended!).
